File: Python/Letter Combinations of a Phone Number.py
"""
Given a digit string, return all possible letter combinations that the number could represent.

A mapping of digit to letters (just like on the telephone buttons) is given below.

Input:Digit string "23"
Output: ["ad", "ae", "af", "bd", "be", "bf", "cd", "ce", "cf"].
"""

class Solution:
    # @return a list of strings, [s1, s2]
    def letterCombinations(self, digits):
        if len(digits) == 0:
            return [""]
        table = {'0':'0','1':'1','2':'abc','3':'def','4':'ghi','5':'jkl','6':'mno','7':'pqrs','8':'tuv','9':'wxyz'}
        result = []
        for i in range(len(digits)):
            if i == 0:
                tmp = table[digits[i]]
                for j in range(len(tmp)):
                    result.append(tmp[j])
            else:
                previous = result
                result = []
                tmp = table[digits[i]]
                for k in range(len(tmp)):
                    for j in range(len(previous)):
                        result.append(previous[j] + tmp[k])
        return result

class Solution:
    def letterCombinations(self, digits):
        lookup, res = ["", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"], []
        self.letterCombinationsRecur(lookup, res, "", digits)
        return res
        
    def letterCombinationsRecur(self, lookup, res, current, digits):
        if digits == "":
            res.append(current)
            return
        for choice in lookup[int(digits[0])]:
            self.letterCombinationsRecur(lookup, res, current + choice, digits[1:])

File: Java/Letter Combinations of a Phone Number.java
/*
Given a digit string, return all possible letter combinations that the number could represent.

A mapping of digit to letters (just like on the telephone buttons) is given below.



Input:Digit string "23"
Output: ["ad", "ae", "af", "bd", "be", "bf", "cd", "ce", "cf"].
Note:
Although the above answer is in lexicographical order, your answer could be in any order you want.
*/

public class Solution {
    private String[] keyboard = new String[]{" ", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
    
    public List<String> letterCombinations(String digits) {
        List<String> result = new ArrayList<>();
        if (digits.length() == 0) return result;
        generateCombinations(result, 0, digits, "");
        return result;
    }
    
    private void generateCombinations(List<String> result, int cur, String digits, String path) {
        if (cur == digits.length()) {
            result.add(path);
            return;
        }
        
        for(int i = 0 ; i < keyboard[digits.charAt(cur) - '0'].length() ; i ++) {
            path += keyboard[digits.charAt(cur) - '0'].charAt(i);
            generateCombinations(result, cur + 1, digits, path);
            path = path.substring(0, path.length() - 1);
        }
    }
}
